Add moved label for topic/stream changes

Subhasish-Behera opened the pull request:
<!-- Please see https://github.com/zulip/zulip-terminal#contributor-guidelines ! -->

**What does this PR do?**  <!-- Overall description goes here -->

<!-- If fixing a filed bug or new feature, add 'Fixes #<issue>' or 'Partial fix for #<issue>' -->
Fixes #1253
<!-- Add a link to a discussion on chat.zulip.org, if relevant -->

**Tested?** <!-- Fine to leave some of these unchecked if this is a draft/work-in-progress -->
- [X] Manually
- [X] Existing tests (adapted, if necessary)
- [x] New tests added (for any new behavior)
- [X] Passed linting & tests (each commit)
<!-- Code must pass CI (GitHub Actions) before merging - look for the green tick! -->

<!-- See https://github.com/zulip/zulip-terminal#commit-style -->
**Commit flow** <!-- if more than one commit; add/delete/fill-in as appropriate -->
<!-- For example:
- first commit doing some thing
- maybe multiple commits doing similar things
-->
commit 1: 
1. add the moved_message field in class Index 
2. Add the logic of when to add message to index["moved_messages"] and index["edited_messages"].

Tried to achieve parity with the logic of Zulip webapp's handling of edits 
mainly from the `analyse_edit_history` function of `zulip/static/js/message_list_view.js` 
(web app's code doesn't use indexing like ZT though)

commit 2:

the second commit focusses on implementing the same logic in case of update_message_event. But in first commit the logic was related to the response values of `Message` but in this case its with `update_message` event.
does similar job.

In both of the commits , the code identifies a real topic change vs unresolve/resolve 
**Notes & Questions** <!-- if any; add/delete/fill-in as appropriate -->
<!-- For example:
- this doesn't include feature X (yet?)
- unsure about Y
- should this do Z?
-->

**Interactions** <!-- if any; add/delete/fill-in as appropriate -->
<!-- eg.
- Waiting on #<PR>
- Blocks #<PR>
-->

Sushmey commented:
Thank you for your effort @Subhasish-Behera !
The code looks well written and seems to work like it's supposed to. Are these all the flags that need checking under message update event? Also, tests normally [don't need to be mentioned](https://github.com/zulip/zulip-terminal#commit-message-style) in the commit title/summary. 
Other than that everything looks good!

srdeotarse commented on tests/helper/test_helper.py:
```diff
@@ -144,7 +144,6 @@ def test_index_edited_message(
     for msg_id, msg in expected_index["messages"].items():
         if msg_id in edited_msgs:
             msg["edit_history"] = []
-
```
if line omission is the only change in the file, then it is not required.

srdeotarse replied:
if line omission is the only change in the file, then it is not required.

srdeotarse commented on zulipterminal/helper.py:
```diff
@@ -47,6 +47,9 @@
 
+RESOLVED_TOPIC_PREFIX = "✔ "
+
+
```
Adding new lines just increases the no. of changes. So, try minimizing your code changes by sticking to only code additions and omissions, not the spacing.

srdeotarse commented on zulipterminal/model.py:
```diff
@@ -64,6 +64,8 @@
 
 
+RESOLVED_TOPIC_PREFIX = "✔ "
+
```
As mentioned earlier about spacing.

srdeotarse replied:
As mentioned earlier about spacing.

srdeotarse commented on zulipterminal/helper.py:
```diff
@@ -399,8 +404,44 @@ def index_messages(messages: List[Message], model: Any, index: Index) -> Index:
+                                RESOLVED_TOPIC_PREFIX
+                            )
+                            and edit_history_event["prev_topic"][2:]
+                            != edit_history_event["topic"][2:]
```
I think you have not handled moved message case if message is edited to resolve/unresolve topic with same name.

srdeotarse commented on zulipterminal/helper.py:
```diff
@@ -47,6 +47,9 @@
 
+RESOLVED_TOPIC_PREFIX = "✔ "
+
+
```
Adding new lines just increases the no. of changes. So, try minimizing your code changes by sticking to only code additions and omissions, not the spacing.

neiljp replied:
Apologies, I should have said that this was already defined in api_types.py - it has to be consistent with the API for resolving to be recognized.

Subhasish-Behera commented:
> Thank you for your effort @Subhasish-Behera ! The code looks well written and seems to work like it's supposed to. Are these all the flags that need checking under message update event? Also, tests normally [don't need to be mentioned](https://github.com/zulip/zulip-terminal#commit-message-style) in the commit title/summary. Other than that everything looks good!

Changed the commit messages. 
Re message update event. ZT treats update_message_event(content/stream/topic changes at least) almost like it's a message event but with a different `response` value. I thought that way and came up with the code for `update_message_event`. I am not solving any other bug/enhancement related to `edit_history` or `update_message_event` here i.e This code only applies the `MOVED` label if in the previous version of these functions  `EDITED` label can be applied to those messages.

srdeotarse commented on zulipterminal/ui_tools/messages.py:
```diff
@@ -723,14 +723,24 @@ def main_view(self) -> List[Any]:
         if self.message["id"] in self.model.index["edited_messages"]:
             edited_label_size = 7
             left_padding = 1
+            is_edited = True
```
Instead of ``is_edited = True``, you can replace it directly with ``label_text = "EDITED`` and remove last conditional to reduce no. of code lines while keeping the code logic intact.
Same can be done with ``is_edited = False``

neiljp replied:
The current changes don't look like they've resolved this issue yet?

I'd also suggest moving this change into its own commit.

srdeotarse reviewed (commented):
Thanks for working on this issue 👍. You have understood the issue and have tried to accommodate all conditions that distinguish ``MOVED`` and ``EDITED`` messages in ZT.

Just a suggestion that it would be really good if you add comments for conditionals related to (un)resolve topics for better understanding.

neiljp commented on zulipterminal/model.py:
```diff
@@ -1591,7 +1593,31 @@ def _handle_update_message_event(self, event: Event) -> None:
 
         if indexed_message:
-            self.index["edited_messages"].add(message_id)
+            if "orig_content" in event:
+                self.index["edited_messages"].add(message_id)
+            if "prev_stream" in event:
+                self.index["moved_messages"].add(message_id)
+            if "subject" in event:
+                if not event["subject"].startswith(RESOLVED_TOPIC_PREFIX):
```
This looks very similar to the logic in the previous commit. Once you've simplified both, are they simple enough to combine into one function?

neiljp reviewed (commented):
@Subhasish-Behera Thanks for looking into this :)

Practically speaking this looks to work OK after a quick test :+1:

However, the logic is rather difficult to read in parts, partly due to reuse of long variables, and I'm not sure how much has come from server or web app code.

It also looks like you could potentially implement the update part using similar functionality to the first commit, though that may only be clear once the code is easier to read.

There are effectively 3 changes here (at least), and while the last commit says tests added, it seems like it's only updated some tests to pass. We really need tests for all the new behavior, and the messages code doesn't require any test updates, which suggests we don't have a pre-existing test either.

If you end up extracting the common logic between the two parts that I pointed out, testing that function may simplify testing the other parts.

Subhasish-Behera commented:
Now, the `analyse_edit_history` is present in https://github.com/zulip/zulip/blob/main/web/src/message_edit_history.js

neiljp commented:
Ready to review? If so, please switch the PR labels :)

Subhasish-Behera commented:
still working on changing the code to be a single function and its tests.

Commit:
api_types: Add edit_history to Message.

Commit:
helper: Add a function analyse_edit_history.

The function analyse_edit_history is called from index
_messages of helper.py. It decouples the edit/moved label
adding mechanism from index_messages to be used later in
_handle_update_messages_event.Stream/content changes comes
under "edited" while topic editing checks cases related to
resolve/unresolve etc which can go either way of "moved"/"edited".
But moved label is applied to specific cases while rest of the cases
are "edited" using else conditional(expect the resolve_change case).

Commit:
messages: Add condition for moved label

Instead of EDITED label now urwid.Text takes label_text as argument which can
be "EDITED" and "MOVED".

Commit:
model: update _handle_update_messages_event.

Uses the analyse_edit_history funciton from helper.py
The parameters passed to analyse_edit_history is slightly
diffrent from earlier as it gets the required values from event
information.
Tests adapted.

Commit:
test_messages: Add and modify test for message labels.

Tests the conditions for moved and edited label along wth
UI differences.
Fixes #1253

Subhasish-Behera commented:
@zulipbot add "PR needs review"

Commit:
test_helper: Added tests for analyse_edit_history.

It tests the possible scenarios for the function as expressed in the test ids.

mounilKshah reviewed (commented):
Thank you for working on this @Subhasish-Behera !

I ran the code on my local machine and it worked as expected. The tests written also seem to cover all the possible cases. I manually tried testing for all the cases and it worked perfectly fine. As for the commit structure, we can discuss it on the ZT stream.

neiljp commented on zulipterminal/helper.py:
```diff
@@ -271,6 +274,44 @@ def set_count(id_list: List[int], controller: Any, new_count: int) -> None:
+    resolved_prefix = RESOLVED_TOPIC_PREFIX + " "
+    if content_changed or stream_changed:
+        index["edited_messages"].add(msg_id)
+    elif old_topic:
```
This also depends upon current_topic being non-None, which as it stands is not guaranteed. In other similar cases we have an assert at the top of the function to ensure that 'coordinated' values are used correctly. The alternative is to define the parameter as a data structure.

neiljp reviewed (commented):
@Subhasish-Behera I intended to only remove the needs-review label, since this needs a rebase and may have outstanding comments to address, including from Mounil here and in the stream. However, I gave this another look and left some further notes.

It's great to see the test(s) :+1: I've not read all of the main function test cases this time, and that test may become simpler if my comments result in adjustment of eg. the function signature (parameters/return/side-effects).

A point I didn't make in an inline comment, is that limiting the UI from directly accessing the model (index) data is something we should aim for. So we could add a model accessor method which returns the 'edit status' as eg. a string (a limited Literal type, perhaps), with that method being the part that interprets the edited/moved sets - and would be easier to test. We could still test the UI, but it'd be a mapping from a string<->UI, so likely less important.

Keeping that interpretation of the sets in the model may make the way in which the sets are updated, eg. whether the sets are kept disjoint, less important - the UI just gets one of 3 results via the accessor method.

I do think it's important that we resolve the current/old topic issue I mention inline, since currently it's unclear what should happen with multiple topic renames in the message history, at least with the current logic.

neiljp commented:
@Subhasish-Behera I've just rebased and pushed a version of this with the commit text summary split by a blank line (which you didn't have early on while contributing) and rebased to skip adding the symbol and import it instead - since that was the main conflict.

I suspect this needs a little rework due to other changes - it didn't quite pass `check-branch` locally. That's partly related to the typing of the `update_message` event, but also a model test failure seems present now.

If you do have updates locally, then feel free to push over this version, but please do fix conflicts and compare it to the changes I just pushed first.

To get this merged, I'd suggest moving the tests into the commit with the related changes, and leaving the update_message commit at the end since that's the update-during-session extension - and the one that may be failing a test now? Some of my points above were comments, but we should certainly look at the server version dependency, and ensure the tests are clear.
